Attached class param docs (#2636)
sorbet/sorbet: website/docs/attached-class.md

Child.make.say_hi # No error now, as Child.make has the type Child
```

## `T.attached_class` as an argument?

At this point, it's natural to think about writing methods that accept arguments
that take values of type `T.attached_class`; you would get more specific type
information about the arguments when calling from a sub-class, so why not?

> Consider this example:

Reviewer: This seems confusing. Maybe just jump to the point?

> Using `T.attached_class` as an argument won't work. To see why, let's look at an example:

Author: Good idea. I was a little worried that this introduction would lead people to believe that it was possible, and that if you just skimmed the docs you'd be confused about the errors you'd get.

```ruby
class Parent
  extend T::Sig

  sig {returns(T.attached_class)}
  def self.make
    new
  end

>   sig {params(x: T.attached_class).void}

Reviewer: Can we add an inline comment to this snippet saying that they should **not** expect this line to work? It makes it easier for people who are not reading and instead just skimming.

  def self.consume(x)

>   end

Reviewer: ```suggestion
    # ...
  end
```

end

class Child < Parent; end

Parent.consume(Parent.new)
Child.consume(Parent.new) # We would like this to be an error
```

However, problems arise when you begin passing around singleton classes. Imagine
that you write the method below, that accepts arguments of type
`T.class_of(Parent)`.

```ruby
class A
  extend T::Sig

  sig {params(cls: T.class_of(Parent)).void}
  def self.consume_parent(cls)
    cls.consume(Parent.make)
  end
end
```

When we pass in `Parent` as the argument, everything is fine: `Parent.make`
returns a value of type `Parent`, and `cls.consume` expects an argument of type
`Parent`. However when we pass `Child` as an argument, problems arise:
`Parent.make` still makes a value of type `Parent`, and since the only thing we
know about `cls` is that it is a subtype of `T.class_of(Parent)`, we are forced
to assume that its `T.attached_class` will be `Parent`. The result of this, is
that the signature for `Child.consume` is violated, as it expects an argument

> whose type is a subtype of `Child`, and `Parent` doesn't satisfy that condition.

Reviewer: This paragraph kind of lost me. Maybe it would be instructive to phrase this in terms of expected / actual types?

> whose type is a subtype of `Child`, and `Parent` doesn't satisfy that condition.

Reviewer: This paragraph kind of lost me. Maybe it would be instructive to phrase this in terms of expected / actual types?

Because of this situation, `T.attached_class` is only allowed to show up in the
`returns` part of a signature, and if it does show up in the `params` of a
signature, you'll get an error:

```ruby
class Parent
  extend T::Sig

  sig {params(x: T.attached_class).void}
            # ^: error: type_template <AttachedClass> was defined as :out but is

>             # used in an :in context

Reviewer: I wonder if we should special-case this error message; then we could link to this doc from that error message directly.

Author: I like that idea, I can see this error message as being confusing without an explanation.

Child.make.say_hi # No error now, as Child.make has the type Child
```

## `T.attached_class` as an argument?

> Using `T.attached_class` as the type of parameters is an error in sorbet, and

Reviewer: ```suggestion
Using `T.attached_class` as a parameter type is an error in Sorbet, and
```

allowing it would cause soundness issues. To see why, let's take a look at an
example:

```ruby
class Parent
  extend T::Sig

  sig {returns(T.attached_class)}
  def self.make
    new
  end

  sig {params(x: T.attached_class).void} # A bad type definition for `x`
  def self.consume(x)
    puts "consumed"
  end
end

class Child < Parent
  extend T::Sig

  sig {void}
  def say_hi
    puts "hi"
  end

  sig {params(x: T.attached_class).void} # A bad type definition for `x`
  def self.consume(x)
    x.say_hi
  end
end

Parent.consume(Parent.new)
Child.consume(Parent.new) # We would like this to be an error
```

However, problems arise when you begin passing around singleton classes. Imagine

> that you write the method below, that accepts arguments of type

Reviewer: ```suggestion
writing the method below, that accepts arguments of type
```

`T.class_of(Parent)`.

```ruby
class A
  extend T::Sig

  sig {params(cls: T.class_of(Parent)).void}
  def self.consume_parent(cls)
    cls.consume(Parent.make)
  end
end
```

For the call to `cls.consume` in the body of `A.consume_parent`, we are always
passing an argument whose type is `Parent`. Let's walk through two different
calls to `A.consume_parent`:

- `A.consume_parent(Parent)`
  - `Parent` has type `T.class_of(Parent)`, and is acceptable for `cls`
  - `Parent.consume` accepts values of type `Parent` (via `T.attached_class`)
- `A.consume_parent(Child)`
  - `Child` has type `T.class_of(Child)`, which is a subtype of
    `T.class_of(Parent)` and is acceptable for `cls`
  - `Child.consume` accepts arguments of type `Child` (via `T.attached_class`),
    and thus will raise an error when `say_hi` is called on the instance
    produced by `Parent.make`

> 

Reviewer: I might say something here to drive the point home like:

> As these two cases show, Sorbet can't know whether the body of `A.consume_parent` type checks or not without knowledge about what type `cls` is **at runtime**. The standard way to prevent ambiguities like this is to prevent `Parent.consume` from being defined using `T.attached_class` in the first place.

Reviewer: (reinforces what we mean by "Because of **this problem**")
